3.next- Start implementing the Arguments class.

markstory opened the pull request:
Build out basic positional argument methods. I've skipped methods to fetch positional arguments by name, as I don't think it makes sense now. The whole point of positional arguments is that they are _positional_.

Refs #11137

markstory committed:
Start implementing the Arguments class.

Build out basic positional argument methods. I've skipped methods to
fetch positional arguments by name, as I don't think it makes sense now.
The whole point of positional arguments is that they are _positional_.

Refs #11137

lorenzo commented on src/Console/Arguments.php:
```diff
@@ -0,0 +1,84 @@
+     * @param array $args Positional arguments
+     * @param array $options Named arguments
+     */
+    public function __construct(array $args, array $options)
```
Should it maybe have the ability to name arguments? like being able to address the 3rd argument by name.
I've found this to be very useful in other frameworks.

dereuromark replied:
I would then use a similar approach of naming as I mentioned in the main ticket:

    getArgumentAt(index)
    hasArgumentAt(index)

and the new

    hasArgument(string)

maybe then.

markstory replied:
@lorenzo Are you thinking of a map between names and indexes?

lorenzo replied:
Yes, exactly

markstory replied:
Ok. I was thinking of skipping that. But if you think it will be valuable to have. I'll add it in.

markstory replied:
I've renamed the methods to include 'At'. I'll add the name based methods in a subsequent pull request.

dereuromark commented on src/Console/Arguments.php:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * Positional arguments.
+     *
+     * @var array
```
Are those actuall `string[]`?

dereuromark commented on src/Console/Arguments.php:
```diff
@@ -0,0 +1,84 @@
+    /**
+     * Get all positional arguments.
+     *
+     * @return array
```
Same here then.

markstory committed:
Rename methods to make room for named positional arguments.

Rename methods to use 'At' suffix to indicate a positional index. New
methods will be added that can read/check positional arguments by
'name'.

markstory committed:
Fix type hints.

markstory committed:
Add hasArgument and getArgument

Allow positional arguments to be accessed by name

markstory commented:
@lorenzo @dereuromark I've added named argument support.

saeideng commented on tests/TestCase/Console/ArgumentsTest.php:
```diff
@@ -57,9 +56,43 @@ public function testGetArgumentAt()
+    public function testGetArgument()
+    {
+        $values = ['big', 'brown', 'bear'];
+        $names = ['size' => 0, 'color' => 1, 'species' => 2, 'odd' => 3];
```
how happen if  `$name` set like bellow 
```php
$names = ['size', 'color', 'species', 'odd'];
//or 
$names = [0 => 'size', 1 => 'color', 2 => 'species',3 =>  'odd'];
```

markstory replied:
@saeideng We'd have to flip that data so we can use `isset()`, or switch to using `array_search()`. Using `array_search()` might be simpler for callers now that you mention it.

robertpustulka commented on src/Console/Arguments.php:
```diff
@@ -0,0 +1,125 @@
+            return $this->args[$index];
+        }
+
+        return null;
```
Maybe the exception should be thrown if an argument is not present?

markstory replied:
Is an argument missing exceptional? The CLI parser should throw errors when required arguments are missing.

robertpustulka commented on src/Console/Arguments.php:
```diff
@@ -0,0 +1,125 @@
+    public function getArgument($name)
+    {
+        if (!isset($this->argNames[$name])) {
+            return null;
```
Same here.

markstory committed:
Make argNames parameter a simple list.

This makes it easier to integrate with the ConsoleOptionParser.
